staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go

```diff
@@ -557,14 +557,22 @@ func (c *Cacher) Watch(ctx context.Context, key string, opts storage.ListOptions
 	//   watchers on our watcher having a processing hiccup
 	chanSize := c.watchCache.suggestedWatchChannelSize(c.indexedTrigger != nil, triggerSupported)
 
+	resourceVersionGuard := newStorageResourceVersionGuard(c.storage, c.newListFunc, c.resourcePrefix, c.objectType.String())
```

Comment:
I think this is now much overcomplicated at this point.

If we want to ensure that watchcache is synced to at least a given RV before we start sending events we simply:
- get the current RV (if RV="" we read it from etcd, of 0 we read it from watchcache)
- do "waitUntilWatchCacheFreshAndForceAllEvents()"
- send bookmark

Most of the logic that we have now is not really needed, and we should really simplify it.